crates/solver/src/solver/naive_solver/multi_order_solver.rs

```diff
@@ -148,18 +148,36 @@ fn solve_with_uniswap(
     // Because the smart contracts round in the favour of the traders, it could
     // be that we actually require a bit more from the Uniswap pool in order to
     // pay out all proceeds. We move the rounding error to the sell token so that
-    // it either comes out of the fees or existing buffers. Compute that amount:
-    let uniswap_out_with_rounding = big_int_to_u256(&settlement.trade_executions().fold(
+    // it either comes out of the fees or existing buffers. It is also important
+    // to use the **synthetic** order amounts for this, as this output amount
+    // needs to be computed for what is actually intented to be traded against
+    // an AMM and not how the order will be executed (including things like
+    // surplus fees).
+    let uniswap_out_with_rounding = big_int_to_u256(&orders.iter().try_fold(
         BigInt::default(),
-        |mut total, execution| {
-            if execution.sell_token == uniswap_out_token {
-                total -= execution.sell_amount.to_big_int();
+        |mut total, order| {
+            if order.sell_token == uniswap_out_token {
+                total -= match order.kind {
+                    OrderKind::Sell => order.sell_amount,
+                    OrderKind::Buy => order
+                        .buy_amount
+                        .checked_mul(uniswap_out)?
+                        .checked_div(uniswap_in)?,
```

Comment:
Naively I would think that the multiplications and divisions in the different cases need to be inverted to each other, right? (line 164/165 and line 172/173).

In the first branch we compute `sell_amount = buy_amount * sell_price / buy_price` with `uniswap_out` as the `sell_price` and `uniswap_in` as the `buy_price`.
In the second branch we compute `buy_amount = sell_amount * buy_price / sell_price`  with `uniswap_out` as the `buy_price` and `uniswap_in` as the `sell_price`.

Shouldn't it be consistent across the 2 branches which uniswap value we use as the `buy_price` and which as the `sell_price`?

Reply:
Unintuitively, prices are `amount ^ -1`:
```rust
    let mut settlement = Settlement::new(maplit::hashmap! {
        uniswap_in_token => uniswap_out,
        uniswap_out_token => uniswap_in,
    });
```

So, if `sell_token == uniswap_out_token`, then
```
buy_amount * uniswap_out / uniswap_in
  = buy_amount * uniswap_in_token_price / uniswap_out_token_price
  = buy_amount * buy_price / sell_price
```

Which I think is correct.

Reply:
But isn't the equation we try to uphold `sell_amount / sell_price == buy_amount / buy_price`? And solved for `sell_amount` it would be `sell_amount = buy_amount * sell_price / buy_amount` which is different from what you end up with.
Not sure where I'm going wrong but I guess if all the existing tests still pass it must be alright. 🤔

Reply:
No, its:
```
sell_amount * sell_price == buy_amount * buy_price
```

Or, this is how I remember it:
```
1 beer * 4.00€ == 8 chocolates * 0.50€
```

Reply:
Yeah, that makes more sense than what I had in mind. 🤦‍♂️